### ISSUE TITLE
`cppcoreguidelines-missing-std-forward` should consider expressions in lambda captures.

### ISSUE THREAD

**Issue by DXPower** [llvm/llvm-project#75891]
The following code triggers a false positive in Clang-Tidy 17 and many versions prior:

```cpp
#include <utility>

template<typename F>
auto MakeFunc(F&& func) {
    return [func2 = std::forward<F>(func)]() {
        func2();
    };
}
```

```
[warning: forwarding reference parameter 'func' is never forwarded inside the function body [cppcoreguidelines-missing-std-forward]
```

However, this is a false positive because the `func` variable is captured by an expression in which `std::forward` is indeed called on it. It's pretty easy to test that `std::forward` is being called too:

```cpp
#include<utility>
template<typename F>auto MakeFunc(F&&func){
    return[func=std::forward<F>(func)]{
        func();
    };
}
#include<iostream>
struct foo{
    foo(){std::cout<<"default construction\n";}
    foo(const foo&){std::cout<<"copy construction\n";}
    foo(foo&&){std::cout<<"move construction\n";}
    foo&operator=(const foo&){std::cout<<"copy assignment\n";return*this;}
    foo&operator=(foo&&){std::cout<<"move assignment\n";return*this;}
    void operator()()const&{std::cout<<"lvalue\n";}
    void operator()()const&&{std::cout<<"rvalue\n";}
};
int main(){
    MakeFunc(foo())();
    foo f;
    MakeFunc(f)();
}
```

Prints:
```
default construction
move construction
lvalue
default construction
copy construction
lvalue
```

`cppcoreguidelines-missing-std-forward` should consider expressions in lambda captures.

Godbolt's for the above snippets:
https://godbolt.org/z/9sh3szr7c
https://godbolt.org/z/sYxhsss14

**Comment by iillyyaa** [comment]
Another form of this false positive is when the parameter is captured by reference, and the forwarding is happening inside the lambda body:
```c++
#include <utility>

template <typename T>
void f(T&& x)
{
    [&x] { T t(std::forward(x)); }();
}
```
https://godbolt.org/z/caWzGnTWb

**Comment by A-x-S** [comment]
Using Clang-Tidy 18 I always get false positive for this check. Even with the simplest template like this (mocking c++20 ranges in c++17), the parameter _pred_ is marked with **cppcoreguidelines-missing-std-forward**:
```
#include <algorithm>

template<typename T, typename Pr>
bool all_of(T const& iterable, Pr&& pred)
{
    return std::all_of(iterable.begin(), iterable.end(), std::forward<Pr>(pred));
}

```

**Comment by danieljennings** [comment]
Has this bug been fixed in trunk? I've run into something similar in a non-trunk clang-tidy (shipped as part of some other developer tools, I don't know off-hand which version it is) but was unable to reproduce on Godbolt which claims to use trunk clang-tidy. Similarly, the Godbolt links in this thread no longer appear to demonstrate the false positives described.
